# In-Class Coding Lab: Data Analysis with Pandas

In this lab, we will perform a data analysis on the **RMS Titanic** passenger list. The RMS Titanic is one of the most famous ocean liners in history. On April 15, 1912 it sank after colliding with an iceberg in the North Atlantic Ocean. To learn more, read here: https://en.wikipedia.org/wiki/RMS_Titanic 

Our goal today is to perform a data analysis on a subset of the passenger list. We're looking for insights as to which types of passengers did and didn't survive. Women? Children? 1st Class Passengers? 3rd class? Etc. 

I'm sure you've heard the expression often said during emergencies: "Women and Children first" Let's explore this data set and find out if that's true!

Before we begin you should read up on what each of the columns mean in the data dictionary. You can find this information on this page: https://www.kaggle.com/c/titanic/data 


## Loading the data set

First we load the dataset into a Pandas `DataFrame` variable. The `sample(10)` method takes a random sample of 10 passengers from the data set.

In [2]:
import pandas as pd
import numpy as np

# this turns off warning messages
import warnings
warnings.filterwarnings('ignore')

passengers = pd.read_csv('CCL-titanic.csv')
passengers.sample(10)

PassengerId  Survived  Pclass  \
617          618         0       3   
832          833         0       3   
348          349         1       3   
844          845         0       3   
520          521         1       1   
249          250         0       2   
768          769         0       3   
566          567         0       3   
525          526         0       3   
28            29         1       3   

                                                Name     Sex   Age  SibSp  \
617  Lobb, Mrs. William Arthur (Cordelia K Stanlick)  female  26.0      1   
832                                   Saad, Mr. Amin    male   NaN      0   
348           Coutts, Master. William Loch "William"    male   3.0      1   
844                              Culumovic, Mr. Jeso    male  17.0      0   
520                            Perreault, Miss. Anne  female  30.0      0   
249                    Carter, Rev. Ernest Courtenay    male  54.0      1   
768                              Moran, Mr. Daniel J    male   NaN      1   
566                             Stoytcheff, Mr. Ilia    male  19.0      0   
525                               Farrell, Mr. James    male  40.5      0   
28                     O'Dwyer, Miss. Ellen "Nellie"  female   NaN      0   

     Parch      Ticket     Fare Cabin Embarked  
617      0   A/5. 3336  16.1000   NaN        S  
832      0        2671   7.2292   NaN        C  
348      1  C.A. 37671  15.9000   NaN        S  
844      0      315090   8.6625   NaN        S  
520      0       12749  93.5000   B73        S  
249      0      244252  26.0000   NaN        S  
768      0      371110  24.1500   NaN        Q  
566      0      349205   7.8958   NaN        S  
525      0      367232   7.7500   NaN        Q  
28       0      330959   7.8792   NaN        Q

## How many survived?

One of the first things we should do is figure out how many of the passengers in this data set survived. Let's start with isolating just the `'Survivied'` column into a series:

In [3]:
passengers['Survived'].sample(10)

286    1
519    0
816    0
110    0
741    0
854    0
60     0
860    0
813    0
785    0
Name: Survived, dtype: int64

There's too many to display so we just display a random sample of 10 passengers. 

- 1 means the passenger survivied
- 0 means the passenger died

What we really want is to count the number of survivors and deaths. We do this by querying the `value_counts()` of the `['Survived']` column, which returns a `Series` of counts, like this:

In [4]:
passengers['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

Only 342 passengers survived, and 549 perished. Let's observe this same data as percentages of the whole. We do this by adding the `normalize=True` named argument to the `value_counts()` method.

In [5]:
passengers['Survived'].value_counts(normalize=True)

0    0.616162
1    0.383838
Name: Survived, dtype: float64

**Just 38% of passengers in this dataset survived.**

### Now you Try it!

**FIRST** Write a Pandas expression to display counts of males and female passengers using the `Sex` variable:

In [6]:
# todo write code here
passengers['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

**NEXT** Write a Pandas expression to display male /female passenger counts as a percentage of the whole number of passengers in the data set.

In [7]:
# todo write code here
passengers['Sex'].value_counts(normalize=True)

male      0.647587
female    0.352413
Name: Sex, dtype: float64

If you got things working, you now know that **35% of passengers were female**.

## Who survivies? Men or Women?

We now know that 35% of the passengers were female, and 65% we male. 

**The next think to think about is how do survivial rates affect these numbers? **

If the ratio is about the same for surviviors only, then we can conclude that your **Sex** did not play a role in your survival on the RMS Titanic. 

Let's find out.

In [8]:
survivors = passengers[passengers['Survived'] ==1]
survivors['PassengerId'].count()

342

Still **342** like we discovered originally. Now let's check the **Sex** split among survivors only:

In [9]:
survivors['Sex'].value_counts()

female    233
male      109
Name: Sex, dtype: int64

WOW! That is a huge difference! But you probably can't see it easily. Let's represent it in a `DataFrame`, so that it's easier to visualize:

In [10]:
sex_all_series = passengers['Sex'].value_counts()
sex_survivor_series = survivors['Sex'].value_counts()

sex_comparision_df = pd.DataFrame({ 'AllPassengers' : sex_all_series, 'Survivors' : sex_survivor_series })
sex_comparision_df['SexSurvivialRate'] = sex_comparision_df['Survivors'] / sex_comparision_df['AllPassengers']
sex_comparision_df

AllPassengers  Survivors  SexSurvivialRate
female            314        233          0.742038
male              577        109          0.188908

 **So, females had a 74% survival rate. Much better than the overall rate of 38%**
 
We should probably briefly explain the code above. 

- The first two lines get a series count of all passengers by Sex (male / female) and count of survivors by sex
- The third line creates DataFrame. Recall a pandas dataframe is just a dict of series. We have two keys 'AllPassengers' and 'Survivors'
- The  fourth line creates a new column in the dataframe which is just the survivors / all passengers to get the rate of survival for that Sex.

## Feature Engineering: Adults and Children

Sometimes the variable we want to analyze is not readily available, but can be created from existing data. This is commonly referred to as **feature engineering**. The name comes from machine learning where we use data called *features* to predict an outcome. 

Let's create a new feature called `'AgeCat'` as follows:

- When **Age** <=18 then 'Child'
- When **Age** >18 then 'Adult'

This is easy to do in pandas. First we create the column and set all values to `np.nan` which means 'Not a number'. This is Pandas way of saying no value. Then we set the values based on the rules we set for the feature.

In [11]:
passengers['AgeCat'] = np.nan # Not a number
passengers['AgeCat'][ passengers['Age'] <=18 ] = 'Child'
passengers['AgeCat'][ passengers['Age'] > 18 ] = 'Adult'
passengers.sample(5)

PassengerId  Survived  Pclass  \
396          397         0       3   
669          670         1       1   
833          834         0       3   
342          343         0       2   
247          248         1       2   

                                                  Name     Sex   Age  SibSp  \
396                                Olsson, Miss. Elina  female  31.0      0   
669  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)  female   NaN      1   
833                             Augustsson, Mr. Albert    male  23.0      0   
342                         Collander, Mr. Erik Gustaf    male  28.0      0   
247                    Hamalainen, Mrs. William (Anna)  female  24.0      0   

     Parch  Ticket     Fare Cabin Embarked AgeCat  
396      0  350407   7.8542   NaN        S  Adult  
669      0   19996  52.0000  C126        S    NaN  
833      0  347468   7.8542   NaN        S  Adult  
342      0  248740  13.0000   NaN        S  Adult  
247      2  250649  14.5000   NaN        S  Adult

Let's get the count and distrubutions of Adults and Children on the passenger list.

In [12]:
passengers['AgeCat'].value_counts()

Adult    575
Child    139
Name: AgeCat, dtype: int64

And here's the percentage as a whole:

In [13]:
passengers['AgeCat'].value_counts(normalize=True)

Adult    0.805322
Child    0.194678
Name: AgeCat, dtype: float64

So close to **80%** of the passengers were adults. Once again let's look at the ratio of `AgeCat` for survivors only. If your age has no bearing of survivial, then the rates should be the same. 

Here's the counts of Adult / Children among the survivors only:

In [14]:
survivors = passengers[passengers['Survived'] ==1]
survivors['AgeCat'].value_counts()

Adult    220
Child     70
Name: AgeCat, dtype: int64

### Now You Try it!

Calculate the `AgeCat` survival rate, similar to how we did for the `SexSurvivalRate`. 

In [15]:
agecat_all_series = passengers['AgeCat'].value_counts()
agecat_survivor_series = survivors['AgeCat'].value_counts()

# todo make a data frame, add AgeCatSurvivialRate column, display dataframe 
agecat_comparision_df = pd.DataFrame({ 'AllPassengers' : agecat_all_series, 'Survivors' : agecat_survivor_series })
agecat_comparision_df['AgecatSurviveRate'] = agecat_comparision_df['Survivors'] / agecat_comparision_df['AllPassengers']
agecat_comparision_df

AllPassengers  Survivors  AgecatSurviveRate
Adult            575        220           0.382609
Child            139         70           0.503597

**So, children had a 50% survival rate, better than the overall rate of 38%**

## So, women and children first?

It looks like the RMS really did have the motto: "Women and Children First."

Here's our insights. We know:

- If you were a passenger, you had a 38% chance of survival.
- If you were a female passenger, you had a 74% chance of survival.
- If you were a child passenger, you had a 50% chance of survival. 


### Now you try it for Passenger Class

Repeat this process for `Pclass` The passenger class variable. Display the survival rates for each passenger class. What does the information tell you about passenger class and survival rates?

I'll give you a hint... "Money Talks"


In [18]:
# todo: repeat the analysis in the previous cell for Pclass 
fly_class_series = passengers['Pclass'].value_counts()
fly_class_survivor_series = survivors['Pclass'].value_counts()

fly_class_comparision_df = pd.DataFrame({'AllPassengers' : fly_class_series, 'Survivors': fly_class_survivor_series})
fly_class_comparision_df['PclassSurvivalRate'] = fly_class_comparision_df['Survivors'] / fly_class_comparision_df['AllPassengers']
fly_class_comparision_df

AllPassengers  Survivors  PclassSurvivalRate
1            216        136            0.629630
2            184         87            0.472826
3            491        119            0.242363